src/services/tau/tau.cpp

```diff
@@ -33,46 +34,76 @@
 /// @file  tau.cpp
 /// @brief Caliper TAU service
 
-#include "../common/ToolWrapper.h"
-#include "caliper/common/filters/RegexFilter.h"
-
+// Caliper annotation bindings for TAU service
 
+#include "caliper/AnnotationBinding.h"
+#include "caliper/common/Attribute.h"
+#include "caliper/common/Variant.h"
 #include <map>
-#include <TAU.h>
+#define TAU_ENABLED
+#define TAU_DOT_H_LESS_HEADERS
+#include "TAU.h"
 
-namespace cali {
 
+using namespace cali;
+
+namespace
+{
+
+class TAUBinding : public cali::AnnotationBinding
+{
+    // when an attribute is created, store the value in TAU as metadata.
+    // this seems reasonable, but might be misguided.
+    static void create_attr_cb(Caliper* c, const Attribute& attr) {
+        std::string tmp("Caliper: " + attr.name());
+        Tau_metadata(tmp.c_str(), c->get(attr).value().to_string().c_str());
+    }
+
+public:
 
-class TAUWrapper : public ToolWrapper {
-  public:
-    virtual void initialize(){
-        TAU_PROFILE_SET_NODE(0);
+    void pre_initialize(Caliper* c) {
+        // register for events of interest
+        c->events().create_attr_evt.connect(&TAUBinding::create_attr_cb);
     }
 
-    virtual std::string service_name() { 
-      return "TAU service";
+    void initialize(Caliper* c) {
+        // initialize TAU
+        int argc = 1;
+        const char *dummy = "Caliper Application";
+        char* argv[1];
+        argv[0] = const_cast<char*>(dummy);
+        Tau_init(argc,argv);
+        // want to get the *real* MPI rank.  How do I get it? like this.
+        Attribute mpi_rank_attr = c->get_attribute("mpi.rank");
+        if (mpi_rank_attr != Attribute::invalid) {
```

Comment:
Should be the other way round.

```suggestion
        if (mpi_rank_attr == Attribute::invalid) {
```